### ISSUE TITLE
Generics

### ISSUE THREAD

**Issue by igorbonadio** [positron-lang/spec#17]
How can we define generic functions and classes?

**Comment by igorbonadio** [comment]
Imagine that any type `T` is an object of type `Type`. So we can define a generic vector class as:

``` ruby
# I will use the syntax I proposed in the issue #9
class Vector(element_type: Type)
  def push_back(element: element_type) -> element_type
    # ...
  end

  def at(index: Int) -> element_type
    # ...
  end
end

vec_of_ints = Vector.new(element_type: Int)
vec_of_ints.push_back(1)
```

In fact, we could pass any expression that evaluates to a `Type` to `(name_var: TypeVar)`.

As I said in the issue #9, "You can read it like the class body is its constructor", so `def`s are expressions too.

But it has a drawback... It can't be done in compile-time (I think).

**Comment by vinivendra** [comment]
Ok. I thought a lot about this and I think I understand it. Personally, though, I have to say I'm not a huge fan.

If we can do this on compile time, I think it might be a good feature to have, though I would still try to clarify the syntax a bit more.

If we become forced to do this on runtime, then I think there may be better alternatives. For instance, `Any`s:

``` ruby
class Vector
    let type: Type

    def push(element: Any) -> void
        assert(element.type == self.type)
        # ...
    end

    def at(index: Int) -> Any
        # ...
    end
end
```

It's not as elegant as generics, but probably more efficient. Honestly, I also think it's more neat and it's **clearer**, but that's coming from someone who's not used to generics.

**Comment by Kazuo256** [comment]
It might be possible to do it at compile time if the C++ strategy is followed. In @igorbonadio's example, it would mean that there is no class `Vector` until an explicit specialization is made in the line below:

``` ruby
vec_of_ints = Vector.new(element_type: Int)
```

And every time the compiler finds a specialization like this, it creates a new version of `Vector` to compile using the given type. Which would mean that the type `Type` is treated as a special case during code parsing. Is that ok?

**Comment by igorbonadio** [comment]
@Kazuo256 the problem is:

``` ruby
def random_type() -> Type
  # return Int or String
end

vec_of_ints = Vector.new(element_type: random_type())
```

In this case we will only know the type at runtime...

Maybe, a solution is to throw an error: "Compiler can't determine element_type".

**Comment by igorbonadio** [comment]
@vinivendra the problem with your solution is that we need cast elements. For example:

``` ruby
let x : Int = vector.at(0)        # error! vector.at() returns an Any
let y : Int = vector.at(0) as Int # ok
```

And we should avoid casts... because, imagine that a vector stores Strings and you cast to Int...

**Comment by Kazuo256** [comment]
Exactly, in the `random_type()` case you would have to forcefully impede the program from compiling or something similar. Which is way I said the `Type` type would have to be treated as a very special case during the compiler's parsing phase in order to detect such situations.

That being the case, having a `Type` type with the same syntax as any other usual type is quite misleading, because they would not really work the same way. Thus, the law of Clarity would be broken. I believe a different approach is needed if supporting generics is desired as a feature for the language.

**Comment by vinivendra** [comment]
@igorbonadio, I think we wouldn't have the risk of receiving the wrong type if the implementation checked the added type every time it added an element, but I agree, it's not a perfect solution.

I think @Kazuo256's idea might be good: treat Type as a different thing, which might be quite ok if we want to enforce static types al the time. Then we can create different Vector classes at compile time.

**Comment by igorbonadio** [comment]
@vinivendra the main problem is not adding elements, but accessing them.

**Comment by vinivendra** [comment]
Yes, but if we guarantee the element is the right type when inserting it, we guarantee it's the right type when accessing.

This is ObjC's option: all Vectors are of `Any`s, and all type handling is a responsibility of the user. Personally, I like C++'s style a lot better: worst case scenario you just create a `Vector<Any>` and get the same results. However, I really think we should strive to do this at compile time, if at all possible.

**Comment by igorbonadio** [comment]
No. When you are accessing an element of type `Int`, you can cast it to `String`. And it is an error. Especially in a Static Typed Language it should not happened.

**Comment by renatocf** [comment]
Marking as opened to #18

**Comment by vinivendra** [comment]
Yeah, I think we're on the same side here. Event with a vector that is guaranteed to have only Integers, the possibility of casting an element to String isnt good.

What do you think about making Types a different thing (not just another type) in order to allow for such generic functions and classes to be created at compile time? If we're really not going to allow dynamic types to happen, having a `Type` class acting as a normal class might not even make that much sense.

**Comment by renatocf** [comment]
I believe we really need to put `Type` as a different thing - one of the parts of the language necessary to do generic metaprogramming. For me, the best way is something similar to C++ templates: a generic is not a type itself, but a factory that produces types when its parameters are created. I'm not sure how Objective-C generics works, but considering what @vinivendra told us, it seem a lot with Java's generics, implemented with type erasure. And I have to say I never liked it: in Java, this feature was introduced to add some type safety to containers that were, in fact, implemented for `Object`'s (Java's superclass-of-all), making sure objects with different subclasses were not put in that container. However, as they didn't changed the language more deeply, the system loosed lots of its power, as a container can't known its type and manipulate it.

Furthermore, I have a question: how do you believe this generic types can be accomplished using hygienic macros? If macros will be our metaprogramming technique (instead of templates, as in C++), I think that generics can be made a special case of macro definitions. For example: as hygienic macros may have parameters, couldn't we use them as generic types? I think it can solve this "type outside of the language". So, what do you think about discussing Hygienic Macros first (I can create an Issue for that) and then discuss more about generics syntax?

**Comment by igorbonadio** [comment]
I think we can have both: generics and hygienic macros. One could be implemented using the other (if it is possible)... but I believe we should have both syntaxes.

As @renatocf said, I think generics should be like a factory that produces types/methods when its parameters are created.

I like the `<T>` syntax:

``` ruby
class Box<T>(value: T)
  def open() -> (content: T)
    content = value
  end
end

var int_box = Box<Int>(123)
var string_box = Box<String>("blah")
# and type can be inferred:
var int_box = Box(123)
var string_box = Box("blah")

class Calc
  def sum<T>(a: T, b: T) -> (result: T)
    result = a + b # as long as T responds to `+`
  end
end

Calc.sum<Int>(1, 2) # => 3
Calc.sum<String>("1", "2") # => "12"
# or
Calc.sum(1, 2) # => 3
Calc.sum("a", "b") # => "ab"
```

**Comment by vinivendra** [comment]
I'm starting to like the idea of generics being factories, though I'm not a fan of the syntax. I don't know, I just really don't like using `<>` as brackets, it seems kind of polluted to me. However, I've been reviewing some things, and I can say that this solution is really preferable to downcasting variables - in fact, I'm thinking about suggesting we disallow downcasting, but that's another discussion.

**Comment by igorbonadio** [comment]
I was thinking about metaprogramming and I think we can implement generics with macros. So what do you think if we discuss about macros before generics?
